keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2909 @@
+general "shape" of the API.
+
+This API separates policy _definition_ from policy _configuration_ by splitting
+responsibilities across resources. The resources involved are:
+
+- Policy definitions (ValidatingAdmissionPolicy)
+- Policy bindings (PolicyBinding)
+- Policy param resources (custom resources)
+
+![Relatinships between policy resources](erd.png)
+
+This allows for a N:N relationship between policy definitions and the configuration of those policies. This separation has already been
+demonstrated successfully by multiple policy frameworks (see the survey further down in this KEP). It has a few key properties:
+
+- Reduces total amount of resource data needed to manage policies:
+  - Params can be shared across multiple policies instead of copied. E.g.
+    multiple policies can be enforce different aspects of a "no external
+    connections", for example, but can all share the configuration.
+  - Policies can be configured in different ways for different use cases without
+    having to copy the policy.
+  - Rollouts and canary-ing can be managed largely via bindings without having
+    to copy policies or params.
+- Ownership of resources aligns well with typical separation of roles for policy
+  management.
+- Existing policy frameworks can leverage this design far more easily because it
+  aligns with how separation of concerns is expressed in most major policy
+  frameworks.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "replicalimit-policy.example.com"
+spec:
+  paramSource:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+  denyReason: Invalid
+  validations:
+    - name: max-replicas
+      expression: "object.spec.replicas <= params.maxReplicas"
+      messageExpression: "'object.spec.replicas must be no greater than ' + string(params.maxReplicas)"
+      # ...other rule related fields here...
+```
+
+The `spec.paramSource` field of the `ValidatingAdmissionPolicy` specifies the
+kind of resources used to parameterize this policy. For this example, it is
+configured by `ReplicaLimit` custom resources. Note in this example how the CEL
+expression references to the parameters via the CEL `params` variable, e.g.
+`params.maxReplicas`.
+
+`spec.matchConstraints` specifies what resources this policy is designed to
+validate. This also guides type-checking, see the "Informational type checking"
+section for details.
+
+`denyReason` assigns a status code the result of this policy if it evaluates to
```

Comment:
This suggests that if you need multiple reasons, you need multiple policies.  A common pattern in baked-in validation is

```
if field is nil {
   return "required" error
} else if field is formatted badly {
   return "invalid" error
}
```

Would it make sense to structure this like:

```yaml
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "replicalimit-policy.example.com"
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  validations:
    - name: field-x-specified
      denyReason: Required
      expression: "object.spec.x != null"
      messageExpression: "'object.spec.x must be provided"
    - name: field-x-valid
      denyReason: Invalid
      expression: "object.spec.x <= 42"
      messageExpression: "'object.spec.x must be greater than 42"
```

That would allow almost transliteration of baked-in validation (furthering the goal of everything being CRDs :)

Reply:
Further refinement.

Built-in validation adds a lot of scaffolding for people to get mostly consistent (and parlty i18n-able) error messages for similar errors.  We do this with `field.Invalid()` and `field.Required()` and so on, which produce distinct structures.  What about something similar here, like:

```yaml
  validations:
    - name: field-x-specified
      expression: "object.spec.x != null"
      failure:
          required:
             fieldPath: "spec.x"
    - name: field-x-valid
      expression: "object.spec.x <= 42"
      failure:
          invalid:
             fieldPath: "spec.x"
             reason: "must be greater than 42"
```

We only need a small number of them (less than 10 in built-in code).  Heck, we could even enumerate common cases like

```yaml
      failure:
          invalid:
             fieldPath: "spec.x"
             greaterThan:
                 comparator:
                     literal: 42
```

Reply:
I see this touched on later.  +1 on more structure and less free-form text, if our goal is to make good, consistent errors.  It takes constant vigilance with the builtin code already.

Reply:
Looks like we agree that `denyReason` should be per-validation.  There was one major KEP update not captured in this PR that fixed it, here's [where things landed](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3488-cel-admission-control/README.md#decisions-and-enforcement):

```yaml
# Policy definition
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: "validate-xyz.example.com"
spec:
  ...
  validations:
    - expression: "self.name.startsWith('xyz-')"
      name: name-prefix
      messageExpression: "self.name + ' must start with xyz-'"
      reason: Unauthorized
    - expression: "self.name.contains('bad')"
      name: bad-name
      message: "name contains 'bad' which is discouraged due to ..."
      code: 400
      reason: Invalid
    - expression: "self.name.contains('suspicious')"
      name: suspicious-name
      messageExpression: "self.name + ' contains suspicious'"
      code: 400
      reason: Invalid
```

Reply:
`fieldPath` is a really good idea. Big +1 for anything that establishes consistent structure.

We mostly punted to alpha2 on how expression failures translate to enforcement actions and how those are all represented. This will be something (the main thing really) that we circle back to this fall as we prep for 1.27 development.

I'll make note (via my next KEP update) of the idea to somehow incorporate the consistent structure of errors that `field.Invalid()` et. al. provide. I like the idea of taking a set of error categories that we've refined over a long time horizon and leveraging them here.